### PR TITLE
Enable automated wheel and dist uploads on releases

### LINKED ISSUE #1300: build binary .whl packages ?
Is there a simple way or a build service that we could use to also build binary wheel (`.whl`) packages (for Linux and OSX), in addition to the conda packages?

- There are Linux build system docker images available at https://github.com/pypa/manylinux
- For OSX we could probably use travis.

### PR DESCRIPTION
**Pull request by IAlibay**
Fixes #1300

Changes made in this Pull Request:
 - Added a deploy action that should build wheels using cibuildwheel, and sdist and then upload them to testpypi / pypi on specific triggers
 - Action runs on every push (so we can notice if something goes wrong), but push to testpypi only happens when you create a tag, and push to pypi only happens when you create a release
 - Created a deploy environment with the necessary pypi / testpypi API token secrets
 - Added maintainer script to normalize release versions in the same way setuptools does (so we can pick it up from testpypi)
 - Standard CI has been expanded to do a full deps check of py3.10 (instead of the minimum deps check done previously)

TODO:
  - Docs

PR Checklist
------------
 - Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: .github/workflows/deploy.yaml
Comment (orbeckst) [inline .github/workflows/deploy.yaml]: If you're using awk (instead of `cut`) then you can do it in one command
```bash
awk '/__version++ =/ {print $3; quit}' package/MDAnalysis/version.py
```
(I *think* the `quit` works.)

**Comment by IAlibay** [comment]
Ok this should be fine now, I've tested the testpypi portion of the release and wheels were built fine here: https://test.pypi.org/project/MDAnalysisTests/2.2.0.dev0/#modal-close

@MDAnalysis/coredevs this needs reviewing as a priority, it's necessary for the 2.2.0 release. It _should_ make our life a lot easier going forward too!

@Seanny123 - I _think_ this should fix the wheel issue, but if you can have a look at if this meets your needs that would be great.

File: .github/workflows/deploy.yaml
Comment (orbeckst) [inline .github/workflows/deploy.yaml:5]: Will this trigger only on branches that start with package- ?

Could you add docs that describe the release workflow to make clearer how it's intended to work?

File: .github/workflows/deploy.yaml
Comment (IAlibay) [reply .github/workflows/deploy.yaml]: done

File: .github/workflows/deploy.yaml
Comment (orbeckst) [inline .github/workflows/deploy.yaml:78]: is `pipx` installed in ubuntu-latest?

File: .github/workflows/deploy.yaml
Comment (orbeckst) [inline .github/workflows/deploy.yaml]: Does this upload wheels as well? Just asking because the name implies "source packages only".

File: .github/workflows/deploy.yaml
Comment (orbeckst) [inline .github/workflows/deploy.yaml:181]: Really cool that we can automatically test the testpypi distribution. Nice.

Comment (orbeckst) [reply .github/workflows/deploy.yaml:181]: How long does it take to complete?

**Review by orbeckst — CHANGES_REQUESTED** [review]
This is great DevOps work, @IAlibay . I have some minor questions because I am not that well versed in all the GH-actions magic and I am happy to positively re-review soon.

The main think I am looking for is minimal maintainer docs that explain the release process and outline the thinking behind the changes, i.e., in plain language, what are the goals and gotchas?

File: .github/workflows/deploy.yaml
Comment (IAlibay) [reply .github/workflows/deploy.yaml:5]: I've updated the wiki notes here: https://github.com/MDAnalysis/mdanalysis/wiki/PreparingReleases

We keep a copy of this in the userguide which I'll update once this PR is merged.

File: .github/workflows/deploy.yaml
Comment (IAlibay) [reply .github/workflows/deploy.yaml:78]: Yup, https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md#package-management

Actually was quite surprising to find, I thought I would have needed to install it, but I guess not.

File: .github/workflows/deploy.yaml
Comment (IAlibay) [reply .github/workflows/deploy.yaml]: Yeah it uploads both, I've updated the names for the actions to reflect this.

File: .github/workflows/deploy.yaml
Comment (IAlibay) [reply .github/workflows/deploy.yaml:181]: ~ 20 mins per runner

It's quite handy that github actions just lets us deploy to lots of runners if the pool is empty, so an aggregate ~ 6 hours set of tests can be squashed down to 20 mins :)

**Comment by IAlibay** [comment]
Ok, on top of updating the release notes in the wiki I've also opened up this PR in the userguide @orbeckst: https://github.com/MDAnalysis/UserGuide/pull/208

I would prefer not have to duplicate docs a third time, so I might suggest that from now on all maintainer docs go in the userguide.

**Comment by Seanny123** [comment]
> @Seanny123 - I _think_ this should fix the wheel issue, but if you can have a look at if this meets your needs that would be great.

This definitely meets my needs. Previously, I was also posting about trying to build the wheels with minimal Numpy versions, but I am no longer concerned about that, even if I can't tell from this PR the compatibility of the wheels being built.

**Comment by IAlibay** [comment]
> > @Seanny123 - I _think_ this should fix the wheel issue, but if you can have a look at if this meets your needs that would be great.
> 
> This definitely meets my needs.

Awesome. Sorry it took so long to get done :( Our release methods are a bit overly complicated.

> Previously, I was also posting about trying to build the wheels with minimal Numpy versions, but I am no longer concerned about that, even if I can't tell from this PR the compatibility of the wheels being built.

So it _should_ be pulling from our [pyproject.toml](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/pyproject.toml) and using the minimum compatible versions of `numpy` for ABI compatibility purposes.

That being said, I agree that it's pretty annoying that it's not clear from the action what's going on. Let me try re-enabling it with a higher verbosity level, maybe that will tell us what it's doing.

**Comment by IAlibay** [comment]
Ok yeah, it is respecting `pyproject.toml` (hence the 1.19.3 pick): https://github.com/MDAnalysis/mdanalysis/runs/6645750506?check_suite_focus=true#step:3:191

I'll leave verbosity at level 1 so we can debug better if something goes wrong :)

**Comment by Seanny123** [comment]
I appreciate the increased verbosity. This will help me onboard other projects of this type (mdtraj in particular) to a CI release pipeline.

>  Sorry it took so long to get done :( Our release methods are a bit overly complicated.

Your methods are perfectly understandable given the constraints! Making Python fast is non-trivial and makes packaging significantly harder. I'm actually super impressed you got it done at all, let alone so quickly!

File: .github/workflows/deploy.yaml
Comment (orbeckst) [reply .github/workflows/deploy.yaml:181]: Nice!

File: .github/workflows/deploy.yaml
Comment (orbeckst) [reply .github/workflows/deploy.yaml:5]: Thank you — I agree with keeping things DRY and to eventually have docs in only one place.

**Review by orbeckst — APPROVED** [review]
Thank you for addressing all my questions and concerns @IAlibay ; I am approving from my end.

However, it would certainly be good if some other @MDAnalysis/coredevs could spare a look.

**Comment by IAlibay** [comment]
> Thank you for addressing all my questions and concerns @IAlibay ; I am approving from my end.
> 
> However, it would certainly be good if some other @MDAnalysis/coredevs could spare a look.

@MDAnalysis/coredevs I'm going to merge this in ~ 24h unless there are any further comments here.

**Review by richardjgowers — APPROVED** [review]

**Pull request merged**
